### ISSUE TITLE
Linear velocity in the observation?

### ISSUE THREAD

**Issue by youzunzhi** [11chens/SLR-master#2]
Congratulations on your amazing work and thank you for sharing this clear and readable code. I have one question regarding the policy observation.

In the paper, you stated that the observation space has 45 dimensions, consisting of no privileged information. However, in the `Go2RoughCfg`, the n_proprio is set to `45+3`:
https://github.com/11chens/SLR-master/blob/8b454a142401ebcf87c33f26bb7b7dba7c0f7e6e/configs/go2_config.py#L38

And, in `compute_observations`, `base_lin_vel` is included.
https://github.com/11chens/SLR-master/blob/8b454a142401ebcf87c33f26bb7b7dba7c0f7e6e/envs/legged_robot.py#L387

I wonder if you can help me understand how the linear velocity is obtained in real deployment. Thanks in advance.

**Comment by 11chens** [comment]
Thank you for reporting this issue. We sincerely apologize for the inconvenience. Due to partial loss of the original source code, errors were introduced during later reimplementation. To address this, you could either:
Remove the linear velocity component entirely, or estimate the linear velocity using the observation history
In our real-world experiments, we did not incorporate linear velocity and observed no significant performance difference.

**Comment by Haku-TS** [comment]
Hi~ @youzunzhi 
please note 
https://github.com/11chens/SLR-master/blob/8b454a142401ebcf87c33f26bb7b7dba7c0f7e6e/modules/actor_critic_slr.py#L54
and 
https://github.com/11chens/SLR-master/blob/8b454a142401ebcf87c33f26bb7b7dba7c0f7e6e/modules/actor_critic_slr.py#L102

Although they included  linear velocity in the buffer, the linear velocity was never input into the network.

**Comment by youzunzhi** [comment]
> Thank you for reporting this issue. We sincerely apologize for the inconvenience. Due to partial loss of the original source code, errors were introduced during later reimplementation. To address this, you could either: Remove the linear velocity component entirely, or estimate the linear velocity using the observation history In our real-world experiments, we did not incorporate linear velocity and observed no significant performance difference.

Thank you very much for your reply. I am sure the trained policy works well without linear velocity. **However, I wonder if you have any plans for releasing your deployment code.** I have spent many hours these days adapting the deployment code from [parkour](https://github.com/ZiwenZhuang/parkour/blob/main/onboard_codes/Deploy-Go1.md) for running the trained policy (that walks well in simulation) on our Unitree A1, but haven't succeeded. It would be extremely helpful if I could use your code to deploy. Could you please let me know if you plan to release the deployment code? Again, thanks for sharing your training code.

**Comment by youzunzhi** [comment]
> Hi~ [@youzunzhi](https://github.com/youzunzhi) please note
> 
> [SLR-master/modules/actor_critic_slr.py](https://github.com/11chens/SLR-master/blob/8b454a142401ebcf87c33f26bb7b7dba7c0f7e6e/modules/actor_critic_slr.py#L54)
> 
> Line 54 in [8b454a1](/11chens/SLR-master/commit/8b454a142401ebcf87c33f26bb7b7dba7c0f7e6e)
> 
>  num_props = num_props-3 # no lin_vel 
> 
> and
> [SLR-master/modules/actor_critic_slr.py](https://github.com/11chens/SLR-master/blob/8b454a142401ebcf87c33f26bb7b7dba7c0f7e6e/modules/actor_critic_slr.py#L102)
> 
> Line 102 in [8b454a1](/11chens/SLR-master/commit/8b454a142401ebcf87c33f26bb7b7dba7c0f7e6e)
> 
>  prop = observations[:, 3:self.num_props+3] 
> Although they included linear velocity in the buffer, the linear velocity was never input into the network.

Thank you for the note. I also noticed this after removing the lin_vel from the observation, so I had to run the training again. 

By the way, I would like to know if you are also trying to deploy the policy on real robots, and have you succeeded yet?

**Comment by Haku-TS** [comment]
> Thank you for the note. I also noticed this after removing the lin_vel from the observation, so I had to run the training again.
> 
> By the way, I would like to know if you are also trying to deploy the policy on real robots, and have you succeeded yet?

Yes, I am trying to deploy the policy an a real go1. I can make it walk, but it doesn't behave as well as in the simulator and as the author demonstrates.

**Comment by youzunzhi** [comment]
> > Thank you for the note. I also noticed this after removing the lin_vel from the observation, so I had to run the training again.
> > By the way, I would like to know if you are also trying to deploy the policy on real robots, and have you succeeded yet?
> 
> Yes, I am trying to deploy the policy an a real go1. I can make it walk, but it doesn't behave as well as in the simulator and as the author demonstrates.

I see. Would you mind sharing your code or at least some reference for the code you are using? I can make Parkour's Go1 policy walk on my A1 using their deployment code, but I can't make this SLR policy walk. I have been debugging for a very long time and tried my best to align the observations and actions.

My email address is youzunzhi@gmail.com. Thank you very much for any help in advance.

**Comment by 11chens** [comment]
> > > Thank you for the note. I also noticed this after removing the lin_vel from the observation, so I had to run the training again.
> > > By the way, I would like to know if you are also trying to deploy the policy on real robots, and have you succeeded yet?
> > 
> > 
> > Yes, I am trying to deploy the policy an a real go1. I can make it walk, but it doesn't behave as well as in the simulator and as the author demonstrates.
> 
> I see. Would you mind sharing your code or at least some reference for the code you are using? I can make Parkour's Go1 policy walk on my A1 using their deployment code, but I can't make this SLR policy walk. I have been debugging for a very long time and tried my best to align the observations and actions.
> 
> My email address is [youzunzhi@gmail.com](mailto:youzunzhi@gmail.com). Thank you very much for any help in advance.

You can refer to the following links for reference: https://github.com/LeCAR-Lab/ABS and https://github.com/ZiwenZhuang/parkour. Please note that the frequency of sending motor commands should be controlled at 5Hz, and the inference frequency should be 20Hz.

**Comment by 11chens** [comment]
> > Thank you for the note. I also noticed this after removing the lin_vel from the observation, so I had to run the training again.
> > By the way, I would like to know if you are also trying to deploy the policy on real robots, and have you succeeded yet?
> 
> Yes, I am trying to deploy the policy an a real go1. I can make it walk, but it doesn't behave as well as in the simulator and as the author demonstrates.

You can refer to the following links for reference: https://github.com/LeCAR-Lab/ABS and https://github.com/ZiwenZhuang/parkour. Please note that the frequency of sending motor commands should be controlled at 5Hz, and the inference frequency should be 20Hz.

**Comment by Haku-TS** [comment]
> > > Thank you for the note. I also noticed this after removing the lin_vel from the observation, so I had to run the training again.
> > > By the way, I would like to know if you are also trying to deploy the policy on real robots, and have you succeeded yet?
> > 
> > 
> > Yes, I am trying to deploy the policy an a real go1. I can make it walk, but it doesn't behave as well as in the simulator and as the author demonstrates.
> 
> I see. Would you mind sharing your code or at least some reference for the code you are using? I can make Parkour's Go1 policy walk on my A1 using their deployment code, but I can't make this SLR policy walk. I have been debugging for a very long time and tried my best to align the observations and actions.
> 
> My email address is [youzunzhi@gmail.com](mailto:youzunzhi@gmail.com). Thank you very much for any help in advance.

I am sorry but I could not share my code. However, I am happy to share my experience:
1, make sure the `dof_pos` and `dof_vel` into the network is in isaac gym order.
2, make sure the previous action into the network is in isaac gym order.
3, make sure the action used to compute torques is in unitree order.
4, do **not** use  `quat_rotate_inverse` for base_ang_vel like:
https://github.com/11chens/SLR-master/blob/8b454a142401ebcf87c33f26bb7b7dba7c0f7e6e/envs/legged_robot.py#L152
because `base_ang_vel` from go1 is already in local coordinate system, while `self.root_states[:, 10:13]` is in world coordinate system.

Here is a piece of my code.

```
obs_curr = torch.cat((base_ang_vel,  # 3, do not use quat_rotate_inverse()
projected_gravity,  # 3`
commands * cmd_scale,  # 3
reindex(dof_pos),  # 12 unitree order ==> isaac order
reindex(dof_vel),  # 12 unitree order ==> isaac order
action_prev), dim=-1).float()  # 12  isaac order, raw from policy output, no clip, no reindex, no scaled
```

**Comment by youzunzhi** [comment]
> > > > Thank you for the note. I also noticed this after removing the lin_vel from the observation, so I had to run the training again.
> > > > By the way, I would like to know if you are also trying to deploy the policy on real robots, and have you succeeded yet?
> > > 
> > > 
> > > Yes, I am trying to deploy the policy an a real go1. I can make it walk, but it doesn't behave as well as in the simulator and as the author demonstrates.
> > 
> > 
> > I see. Would you mind sharing your code or at least some reference for the code you are using? I can make Parkour's Go1 policy walk on my A1 using their deployment code, but I can't make this SLR policy walk. I have been debugging for a very long time and tried my best to align the observations and actions.
> > My email address is [youzunzhi@gmail.com](mailto:youzunzhi@gmail.com). Thank you very much for any help in advance.
> 
> You can refer to the following links for reference: https://github.com/LeCAR-Lab/ABS and https://github.com/ZiwenZhuang/parkour. Please note that the frequency of sending motor commands should be controlled at 5Hz, and the inference frequency should be 20Hz.

Thank you very much for your kind reply and the helpful links.

I'm having a bit of trouble following the suggestion about setting the motor command frequency to 5Hz. When I try that, the robot can't even stay standing. I'm currently using the deployment code from https://github.com/ZiwenZhuang/parkour, and with that, I’m able to make the released Go1 policy walk on my A1. This code uses 50Hz for both motor commands and policy inference, which matches the simulation setup.

When I try using rospy.Rate(5), the robot doesn’t seem to function properly—it can’t stand or walk. So I’m wondering if I might be missing something important in my understanding or code setup. I would really appreciate any further clarification or advice you can share. Thanks again in advance!

**Comment by 11chens** [comment]
> > > > > Thank you for the note. I also noticed this after removing the lin_vel from the observation, so I had to run the training again.
> > > > > By the way, I would like to know if you are also trying to deploy the policy on real robots, and have you succeeded yet?
> > > > 
> > > > 
> > > > Yes, I am trying to deploy the policy an a real go1. I can make it walk, but it doesn't behave as well as in the simulator and as the author demonstrates.
> > > 
> > > 
> > > I see. Would you mind sharing your code or at least some reference for the code you are using? I can make Parkour's Go1 policy walk on my A1 using their deployment code, but I can't make this SLR policy walk. I have been debugging for a very long time and tried my best to align the observations and actions.
> > > My email address is [youzunzhi@gmail.com](mailto:youzunzhi@gmail.com). Thank you very much for any help in advance.
> > 
> > 
> > You can refer to the following links for reference: https://github.com/LeCAR-Lab/ABS and https://github.com/ZiwenZhuang/parkour. Please note that the frequency of sending motor commands should be controlled at 5Hz, and the inference frequency should be 20Hz.
> 
> Thank you very much for your kind reply and the helpful links.
> 
> I'm having a bit of trouble following the suggestion about setting the motor command frequency to 5Hz. When I try that, the robot can't even stay standing. I'm currently using the deployment code from https://github.com/ZiwenZhuang/parkour, and with that, I’m able to make the released Go1 policy walk on my A1. This code uses 50Hz for both motor commands and policy inference, which matches the simulation setup.
> 
> When I try using rospy.Rate(5), the robot doesn’t seem to function properly—it can’t stand or walk. So I’m wondering if I might be missing something important in my understanding or code setup. I would really appreciate any further clarification or advice you can share. Thanks again in advance!

You can refer to the ABS implementation for deployment code, using Numpy+onnx (instead of PyTorch+tensor). The pseudocode is:

action = model_infer()  # 20ms  
for i in range(4):  
    cmd_write_action()  # 5ms  (not 5 Hz)

This ensures high-frequency motor commands while running policy inference at a lower frequency.

**Comment by Wangzeyu6** [comment]
> > > > 谢谢你的提醒。我在从观察中移除 lin_vel 后也注意到了这个问题，所以我不得不重新进行训练。
> > > > 顺便问一下，我想知道你是否也尝试在真实机器人上部署这个策略，并且成功了吗？
> > > 
> > > 
> > > 是的，我正在尝试将策略部署到真正的 go1 中。我可以让它运行，但它的表现不如在模拟器中以及作者演示的那样好。
> > 
> > 
> > 我明白了。您介意分享一下您的代码吗？或者至少提供一些您正在使用的代码的参考资料？我可以使用 Parkour 的 Go1 策略在我的 A1 上用他们的部署代码进行行走，但我无法进行这个 SLR 策略行走。我已经调试了很长时间，并尽力使观察结果和操作保持一致。
> > 我的邮箱是[youzunzhi@gmail.com](mailto:youzunzhi@gmail.com)。非常感谢您的帮助。
> 
> 很抱歉，我无法分享我的代码。不过，我很乐意分享我的经验： 1、确保进入网络的`dof_pos`和`dof_vel`符合 isaac gym 的顺序。2 、确保进入网络的先前操作符合 isaac gym 的顺序。3 、确保用于计算扭矩的操作符合单树顺序。4 、**不要**`quat_rotate_inverse`将base_ang_vel用作 以下函数：
> 
> [SLR-master/envs/legged_robot.py](https://github.com/11chens/SLR-master/blob/8b454a142401ebcf87c33f26bb7b7dba7c0f7e6e/envs/legged_robot.py#L152)
> 
> 第 152 行 [8b454a1](/11chens/SLR-master/commit/8b454a142401ebcf87c33f26bb7b7dba7c0f7e6e)
> 
>  self.base_ang_vel = quat_rotate_inverse ( self.base_quat , self.root_states [ :, 10 : 13 ] )​ ​​​ 
> 
> 因为`base_ang_vel`从go1开始已经在本地坐标系中，而而`self.root_states[:, 10:13]`在世界坐标系中。
> 这是我的一部分代码。
> 
> ```
> obs_curr = torch.cat((base_ang_vel,  # 3, do not use quat_rotate_inverse()
> projected_gravity,  # 3`
> commands * cmd_scale,  # 3
> reindex(dof_pos),  # 12 unitree order ==> isaac order
> reindex(dof_vel),  # 12 unitree order ==> isaac order
> action_prev), dim=-1).float()  # 12  isaac order, raw from policy output, no clip, no reindex, no scaled
> ```
Hello, I don't quite understand the fourth point you proposed: do not use quat_rotate_inverse for base_ang_vel like; What does "starting from go1" mean by local coordinates? In simulation environments, isn't it necessary to convert the angular velocity of the senior coordinate system to one's own coordinate system? If taken directly, it is the angular velocity in the senior sister's coordinate system
